### PR TITLE
added half support

### LINKED ISSUE #3: Support f16, f128
numpy has `float16` and `float128` datatypes, which can appear in npy files.

There exists a crate called `half` for binary16-encoded floats.

* [x] Support f16 (thanks @KeKsBoTer)
* [ ] Support f128

### PR DESCRIPTION
**Pull request by KeKsBoTer**
Hey,

Thanks for the nice and well-built library!
I needed f16/half support, so I added it.

I added some additional tests and all (+wasm) tests pass.
Please let me know if I should add more tests.

What formatter/settings are you using? I used `rustfmt` which cased  a lot of changes... 

Best
Simon

Closes #3

### CODE REVIEW

**Comment by ExpHP** [comment]
Hi! Thank you for the contribution.  I would be glad to have support for f16.

Hmmmm.  From what I recall I don't think I use a formatter.  My own formatting rules are similar to rustfmt defaults (all identation is block-style rather than aligning, and I use trailing commas religiously, even on braced match branches). Can you please try to isolate the meaningful changes from this commit?  Or prefix the current commit with another one that just reformats the current codebase.

**Comment by KeKsBoTer** [comment]
Hey, thanks for the quick reply!
I added a prefix commit just for formatting :) I used the `rustfmt` version v1.6.0.
I just didn't bother with a feature branch due to laziness...sorry 😅

**Comment by ExpHP** [comment]
> I just didn't bother with a feature branch due to laziness...sorry

Ah, okay.  Mostly when I see people do things on master, it's people who haven't done a lot of open source dev.  I always make a feature branch even for the simplest pull request, because oftentimes making a pull request causes the maintainer to notice some trivial mistake somewhere in their docs and then just like that, your master branches have now diverged. 😅 

Thanks for adding the reformat commit!  I'll take a look when I have a minute.

**Comment by ExpHP** [comment]
So I ended up making last minute plans this last weekend that left me largely in no position to review the code (on airplanes and with people).  Thank you for your patience so far!  I should hopefully be able to get to this today.

File: src/write.rs
Comment (ExpHP) [inline src/write.rs:747]: gaaaaaaaaahhhhhhh, this really hurts the readibility.

Hmmm maybe we can force rustfmt to leave it alone with a line comment:

```rust
                std::mem::transmute::<
                    Box<dyn WriteSeek<W> + '_>, // this lifetime satisfies W: '_
                    Box<dyn WriteSeek<W> + 'static>,
                >(Box::new(w))
```

**Comment by ExpHP** [comment]
Okay.  I did a review of the reformat commit to see if there were any issues I had with rustfmt.  Overall I'm fairly satisfied, most of the changes are harmless.  I also didn't realize my editor was not configured to force a newline at EOF!  tsk.

---

**I suggest making a separate PR with the formatting change** now because there's some additional cleanup I'd like with it, and some rustfmt settings I'd like to change.

(after we merge that PR you should be able to rebase this branch on it (or cherry pick the f16 stuff onto a new branch))

---

I would like you to try the following rustfmt config:

**rustfmt.toml**
```rust
match_arm_leading_pipes = "Preserve"
match_block_trailing_comma = true
merge_derives = false
newline_style = "Unix"

# (group_imports = "StdExternalCrate" would be nice but it's unstable)
reorder_imports = false
reorder_modules = false
```

The review I'm about to post contains a bunch of small suggestions on additional fixes.

* Mostly I am bothered by rustfmt's decision to change macro invocations between block-style and inline even though it's too afraid to add/remove trailing commas to them.
  * I made suggestions in the review pane for some of these, but there's WAY too many.

* It did this to attributes too, even system ones (most notably `cfg_attr`).

* Seems this crate was written before `format!("{a} {b}")` syntax was added. That can be used to clean up some string formatting macros that got ugly after reformatting.

* I am not a huge fan of what it did to the tests in `slice.rs`, but... sigh... I won't block this on it.

**Review by ExpHP — COMMENTED** [review]
a large smattering of small comments on rustfmt

File: tests/roundtrip.rs
Comment (ExpHP) [inline tests/roundtrip.rs:8]: - [ ] I think I generally avoid putting `cfg` in tests, especially integration tests because I worry that I could forget to ever actually test it with the feature enabled.
  - You could add `half` to the `required-features` in Cargo.toml for the test.  (doing that could be problematic if `half` was not available on all platforms, but judging from its readme I think it is.)
  - Alternatively, `half` could be tested in a separate test that requires the feature.

**Review by ExpHP — CHANGES_REQUESTED** [review]
Review of the f16 stuff.

This looks very solid!  I only have a small number of comments.

**Review by ExpHP — CHANGES_REQUESTED** [review]
One additional thing not on any of these files:

- [ ] The `Floats` section in [`type_matchup_docs`](https://docs.rs/npyz/latest/npyz/type_matchup_docs/index.html) should be updated.  You can use language similar to what's already there for Complex.

File: Cargo.toml
Comment (ExpHP) [reply Cargo.toml:31]: Wait, lmao actually I'm confused, the public dep links in lib.rs are already guaranteed to be in sync because I'm using ``[`those::cool_ass::smart_links`]``... 🤔

...that comment might be old. 😅

**Pull request closed without merging**
